/**
 * Copyright 2024 AntGroup CO., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 */

syntax = "proto3";
package meta;

message VertexPropertyIndex {
    string name        = 1;
    bool is_unique     = 2;
    string label       = 3;
    uint32 label_id    = 4;
    string property    = 5;
    uint32 property_id = 6;
    uint32 index_id    = 7;
}

message VertexFullTextIndex {
    string name                  = 1;
    string path                  = 2;
    repeated string labels       = 3;
    repeated string properties   = 4;
    repeated uint32 label_ids    = 5;
    repeated uint32 property_ids = 6;
    uint32 index_id              = 7;
}

enum VectorDistanceType {
    L2 = 0; // Euclidean distance
    IP = 1; // Inner product
    COSINE = 2; // Cosine angle, smaller is more similar
}

enum VectorIndexType {
    HNSW = 0;       // Hierarchical navigable small world
    IVF_FLAT = 1;   // InVerted File Flat
}

message VertexVectorIndex {
    string name        = 1;
    string label       = 2;
    uint32 label_id    = 3;
    string property    = 4;
    uint32 property_id = 5;
    uint32 dimensions  = 6;
    VectorIndexType index_type = 7;
    VectorDistanceType distance_type = 8;
    uint32 hnsw_m               = 9;
    uint32 hnsw_ef_construction = 10;
    uint32 index_id             = 11;
    string path                 = 12;
}

message VectorIndexManifest {
    repeated string chunk_ids = 1;
}

message GraphDBMetaInfo {
    string graph_name = 1;
    uint32 graph_id   = 2;
}

enum UpdateType {
    Add = 0;
    Delete = 1;
};

message FullTextIndexUpdate {
    UpdateType type         = 1;
    int64 vid               = 2;
    repeated string fields = 3;
    repeated string values = 4;
}

message VectorIndexUpdate {
    UpdateType type        = 1;
    int64 vid              = 2;
    int64 vector_id        = 3;
    repeated float vector  = 4;
}

message VectorIndexDelta {
    UpdateType type     = 1;
    repeated int64 ids  = 2;
}
